### DISCUSSION TITLE
Best practices on managing event schemas?

### DISCUSSION THREAD

**Discussion by elsurudo** [RailsEventStore/rails_event_store#1767] [General]
Being a relational guy at heart, I'm a bit uneasy about the fact that `data` and `metadata` are basically JSON blobs by default.

I wasn't able to find any best practices in the docs around managing/validating that event schemas are actually valid. Is this a problem I should worry about? Any tips?

Also, what about event versioning? If my schema changes, aren't I going to run into trouble eventually when reading older events?

**Comment by pjurewicz** [comment]
Hello @elsurudo, nice to meet you!

There are many ways to implement events schema validation. It's out of scope of Rails Event Store itself.
However, an example of `dry-types` based event schema is presented in the [`ecommerce`, our demo app](https://github.com/RailsEventStore/ecommerce/blob/master/infra/lib/infra/event.rb)
Another one implementation, using`classy_hash`, has been described in [Domain-Driven Rails](https://blog.arkency.com/domain-driven-rails/) book.

When it comes to versioning and schema changes, please check out these 2 blogposts:
https://blog.arkency.com/4-strategies-when-you-need-to-change-a-published-event/
https://blog.arkency.com/upcasting-events-in-railseventstore/
I believe they cover this topic in details.

**Reply by elsurudo** [reply]
Hi @pjurewicz, nice to meet you too and thanks for the answer :)

In the meantime I've found the exact file in the ecommerce sample and used that as an example, since I was thinking of `Dry::Struct` myself for this.

And thanks for the tips about event changes. I've heard of upcasting and while it makes sense in theory, I can't help but imagine an app accumulating a bunch of cruft over time due to this. But perhaps this is a growing pain I need to get over to get more used to this paradigm?
